flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/physical/stream/WatermarkAssignerChangelogNormalizeTransposeRuleTest.scala

```diff
@@ -0,0 +1,143 @@
+         |)
+         |""".stripMargin)
+  }
+
+  // ----------------------------------------------------------------------------------------
+  // Tests for queries matches WITHOUT_CALC patten
+  // Rewrite always happens in the case
+  // ----------------------------------------------------------------------------------------
+  @Test
+  def testPushdownWatermarkWithoutCalc(): Unit = {
+    val sql =
+      """
+        |SELECT
+        |currency,
+        |COUNT(1) AS cnt,
+        |TUMBLE_START(currency_time, INTERVAL '5' SECOND) as w_start,
+        |TUMBLE_END(currency_time, INTERVAL '5' SECOND) as w_end
+        |FROM simple_src
+        |GROUP BY currency, TUMBLE(currency_time, INTERVAL '5' SECOND)
+        |""".stripMargin
+    util.verifyRelPlan(sql, ExplainDetail.CHANGELOG_MODE)
+  }
+
+  // ----------------------------------------------------------------------------------------
+  // Tests for queries matches WITH_CALC patten
+  // ----------------------------------------------------------------------------------------
+
+  /** push down calc and watermark assigner as a whole if shuffle keys are kept after Calc. */
+  @Test
+  def testPushdownCalcAndWatermarkAssignerWithCalc(): Unit = {
+    val sql =
+      """
+        |SELECT
+        |currency,
+        |COUNT(1) AS cnt,
+        |TUMBLE_START(currency_time, INTERVAL '5' SECOND) as w_start,
+        |TUMBLE_END(currency_time, INTERVAL '5' SECOND) as w_end
+        |FROM src_with_computed_column
+        |GROUP BY currency, TUMBLE(currency_time, INTERVAL '5' SECOND)
+        |""".stripMargin
+    util.verifyRelPlan(sql, ExplainDetail.CHANGELOG_MODE)
+  }
+
+  /** only push down watermark assigner if satisfy all the following condition:
+   *  1. shuffle keys are not kept after Calc
+   *  2. row time field does not depend on computed column
+   */
+  @Test
+  def testPushdownWatermarkAssignerWithCalc(): Unit = {
+    val sql =
+      """
+        |SELECT
+        |TUMBLE_START(currency_time, INTERVAL '5' SECOND) as w_start,
+        |TUMBLE_END(currency_time, INTERVAL '5' SECOND) as w_end,
+        |MAX(rate) AS max_rate
+        |FROM simple_src
+        |GROUP BY TUMBLE(currency_time, INTERVAL '5' SECOND)
+        |""".stripMargin
+    util.verifyRelPlan(sql, ExplainDetail.CHANGELOG_MODE)
+  }
+
+  /**
+   * rewrites has no effect if does not satisfy any of the following condition:
+   * 1. shuffle keys are all kept after Calc
+   * 2. row time field does not depend on computed column
+    */
+  @Test
+  def testNothingHappenWithCalc(): Unit = {
+    val sql =
+      """
+        |SELECT
+        |TUMBLE_START(currency_time, INTERVAL '5' SECOND) as w_start,
+        |TUMBLE_END(currency_time, INTERVAL '5' SECOND) as w_end,
+        |MAX(rate) AS max_rate
+        |FROM src_with_computed_column
+        |GROUP BY TUMBLE(currency_time, INTERVAL '5' SECOND)
+        |""".stripMargin
+    util.verifyRelPlan(sql, ExplainDetail.CHANGELOG_MODE)
+  }
+}
```

Comment:
It's better to add a test which group key is from computed column, such as

` util.addTable(
      s"""
         |CREATE TABLE src_with_computed_column2 (
         |  currency int,
         |  currency2 as currency + 2,
         |  currency_no STRING,
         |  rate  BIGINT,
         |  c STRING,
         |  currency_time as to_timestamp(c),
         |  WATERMARK FOR currency_time AS currency_time - interval '5' SECOND,
         |  PRIMARY KEY(currency) NOT ENFORCED
         |) WITH (
         |  'connector' = 'values',
         |  'changelog-mode' = 'UA,D',
         |  'enable-watermark-push-down' = 'true'
         |)
         |""".stripMargin)

 val sql =
      """
        |SELECT
        |currency2,
        |COUNT(1) AS cnt,
        |TUMBLE_START(currency_time, INTERVAL '5' SECOND) as w_start,
        |TUMBLE_END(currency_time, INTERVAL '5' SECOND) as w_end
        |FROM src_with_computed_column2
        |GROUP BY currency2, TUMBLE(currency_time, INTERVAL '5' SECOND)
        |""".stripMargin
    util.verifyRelPlan(sql, ExplainDetail.CHANGELOG_MODE)
`

Reply:
good point